File: life/src/main/scala/com/github/rd/life/LifeModel.scala
package com.github.rd.life

/*object LifeModel extends App {
  val model = new LifeModel(3, 3)
  model.setCell(1, 0, 1)
  model.setCell(1, 1, 1)
  model.setCell(1, 2, 1)
  printModel

  model.simulate
  printModel
  model.simulate
  printModel
  model.simulate
  printModel

  model.clear
  printModel

  def printModel {
    for (i <- 0 until 3) {
      for (j <- 0 until 3) {
        print(model.getCell(i, j))
      }
      println
    }
    println
  }
}*/

class LifeModel(height: Int, width: Int) {

  private var mainField = Array.fill[Byte](height, width)(0)
  private var backField = Array.fill[Byte](height, width)(0)
  private val neighborsOffsets = Array(
    Array(-1, -1),
    Array(-1, 0),
    Array(-1, 1),
    Array(0, -1),
    Array(0, 1),
    Array(1, -1),
    Array(1, 0),
    Array(1, 1)
  )
  
  def getHeight() = height
  def getWidth() = width

  def setCell(x: Int, y: Int, value: Byte) {
    mainField(x)(y) = value
  }

  def getCell(x: Int, y: Int): Byte = mainField(x)(y)

  def simulate {
    for (x <- 0 until height) {
      for (y <- 0 until width) {
        val neighborsCount = countNeighbors(x, y)
        backField(x)(y) = simulateCell(mainField(x)(y), neighborsCount)
      }
    }
    mainField = for (line <- backField) yield {
      for (elem <- line) yield elem
    }
    backField = Array.fill[Byte](height, width)(0)
  }

  def clear() {
    mainField = Array.fill[Byte](height, width)(0)
    backField = Array.fill[Byte](height, width)(0)
  }
  
  private def countNeighbors(x: Int, y: Int): Int = {
    //def normalizeXY(value: Int, bound: Int): Int = {
    //  if (value == bound) 0
    //  else if (value < 0) bound - 1
    //  else value
    //}
    var count = 0

    for (offsets <- neighborsOffsets) {
      //val xOffset = normalizeXY(x + offsets(0), height)
      //val yOffset = normalizeXY(y + offsets(1), width)
      //count += mainField(xOffset)(yOffset)
      
      val xOffset = x + offsets(0)
      val yOffset = y + offsets(1)
      if (xOffset >= 0 && xOffset < height && yOffset >= 0 && yOffset < width) count += mainField(xOffset)(yOffset)
    }
    count
  }

  private def simulateCell(state: Byte, neighbors: Int): Byte = {
    if (state == 0) if (neighbors == 3) 1 else 0
    else if (neighbors < 2 || neighbors > 3) 0 else 1
  }
}

File: cpp-life/LifeModel.cpp
#include "LifeModel.h"

int LifeModel::getCell(int row, int col)
{
	return frontField[row][col];
}

void LifeModel::setCellOn(int row, int col)
{
	setCell(row, col, 1);
}

void LifeModel::setCellOff(int row, int col)
{
	setCell(row, col, 0);
}

void LifeModel::setCell(int row, int col, int val)
{
	frontField[row][col] = val;
}

void LifeModel::simulate()
{
	for (int i = 0; i != n; ++i)
	{
		for (int j = 0; j != m; ++j)
		{
			int neighboursCount = countNeighbours(i, j);
			int cellState = getCell(i, j);

			if ((cellState == 0 && neighboursCount == 3)
					|| (cellState > 0 && (neighboursCount == 2 || neighboursCount == 3)))
			{
				backField[i][j] = 1;
			}
			else
			{
				backField[i][j] = 0;
			}
		}
	}
	swapFields();
}

int LifeModel::countNeighbours(int row, int col)
{
	static const int offsetsCount = 8;
	static const int offsets[offsetsCount][2] = 
	{
		{-1, -1},
		{-1, 0},
		{-1, 1},
		{0, -1},
		{0, 1},
		{1, -1},
		{1, 0},
		{1, 1}
	};
	int count = 0;

	for (int k = 0; k != offsetsCount; ++k)
	{
		int i = row + offsets[k][0];
		int j = col + offsets[k][1];

		if (i >= 0 && i < n && j >= 0 && j < m)
		{
			count += getCell(i, j);
		}
	}
	return count;
}

void LifeModel::swapFields()
{
	int** temp = frontField;
	frontField = backField;
	backField = temp;
}

void LifeModel::clear()
{
	for (int i = 0; i != n; ++i)
	{
		for (int j = 0; j != m; ++j)
		{
			frontField[i][j] = 0;
			backField[i][j] = 0;
		}
	}
}
